woocommerce/woocommerce-gutenberg-products-block
# Use Cypress for C&C related e2e tests

In #5700, we started identifying critical flows and covering them with e2e tests. After we hardly made any progress with four engineers in three weeks, I decided putting Cypress to the test. While this is in contradiction to WooCommerce, WordPress and Gutenberg, which all use Jest and Puppeteer, I decided trying this anyway. While putting Cypress to the test, I noticed that tests can be written very fast, debugging works like charm, and flaky tests might become history.

### e2e environment

The current Jest/Puppeteer tests set up before and tear down the e2e environment for every test run. This is rather time-consuming when creating tests. I tried a different approach by decoupling the setup and tear down process from the tests itself. The following commands set up the e2e environment, run the tests and tear down the e2e environment:

```shell
// Set up the e2e environment
$ npm run wp-env:config-cypress

// Run the e2e test suite
$ npm run test:e2e-cypress

// Tear down the e2e environment
$ npm run wp-env clean
```

## Integration in GitHub Actions

This PR does not contain a GitHub Action integration yet. This is something which is still on my list.

### Intention

By putting Cypress to the test, it's not my intention to replace Jest and Puppeteer. I intend to improve the way we create e2e tests for the cart and checkout blocks. I created this PR in a way that both Jest/Puppeteer and Cypress can work in parallel.

### Testing

1. Check out this PR.
2. Start Docker Desktop.
3. Run `npm i` to install all dependencies.
4. Run `npm run wp-env clean && npm run wp-env:config-cypress` to set up a fresh e2e environment.
5. Run `npm run test:e2e-cypress` to run the Cypress test suite in the terminal.
6. Run `npm run test:e2e-cypress-dev` to run the Cypress test suite in the Cypress GUI.
